x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -102,15 +103,78 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
-  async createOrUpdateWriteTarget({ namespace }: { namespace?: string }) {
+  async rolloverAliasIfNeeded({ namespace }: { namespace?: string }) {
+    const clusterClient = await this.getClusterClient();
+    const alias = getNamespacedAlias({ alias: this.options.alias, namespace });
+    let simulatedRollover: IndicesRolloverResponse;
+    try {
+      ({ body: simulatedRollover } = await clusterClient.indices.rollover({
+        alias,
+        dry_run: true,
+      }));
+    } catch (err) {
+      if (err?.meta?.body?.error?.type !== 'index_not_found_exception') {
+        throw err;
+      }
+      return;
+    }
+
+    const [writeIndexMapping, simulatedIndexMapping] = await Promise.all([
+      clusterClient.indices.get({
+        index: simulatedRollover.old_index,
+      }),
+      clusterClient.indices.simulateIndexTemplate({
+        name: simulatedRollover.new_index,
+      }),
+    ]);
+    const currentVersions = get(writeIndexMapping, [
+      'body',
+      simulatedRollover.old_index,
+      'mappings',
+      '_meta',
+      'versions',
+    ]);
+    const targetVersions = get(simulatedIndexMapping, [
+      'body',
+      'template',
+      'mappings',
+      '_meta',
+      'versions',
+    ]);
+    const componentTemplateRemoved =
+      Object.keys(currentVersions).length > Object.keys(targetVersions).length;
+    const componentTemplateUpdated = Object.entries(targetVersions).reduce(
+      (acc, [templateName, targetTemplateVersion]) => {
+        const currentTemplateVersion = get(currentVersions, templateName);
+        const templateUpdated =
+          currentTemplateVersion == null || currentTemplateVersion < Number(targetTemplateVersion);
+        return acc || templateUpdated;
+      },
+      false
+    );
+    const needRollover = componentTemplateRemoved || componentTemplateUpdated;
+    if (needRollover) {
+      try {
+        await clusterClient.indices.rollover({
+          alias,
+          new_index: simulatedRollover.new_index,
+        });
+      } catch (err) {
+        if (err?.meta?.body?.error?.type !== 'resource_already_exists_exception') {
+          throw err;
+        }
```

Comment:
Just curious, is there any doc describing exceptions from different methods of `ElasticsearchClient`?

Reply:
I'm not aware of any...it would be amazing if we had one though.